smart-contract-verifier/src/solidity/verifier.rs

```diff
@@ -424,116 +51,245 @@ impl Verifier {
+            abi.constructor(),
+        )?;
 
         Ok((abi.into_owned(), constructor_args))
     }
 
-    /// Checks that solc versions obtained from metadata hash correspond
-    /// for provided deployed bytecode and deployed bytecode obtained
-    /// as a result of local compilation.
-    fn check_metadata_hash_solc_versions(
-        &self,
-        deployed_bytecode: &DeployedBytecode,
-    ) -> Result<(), VerificationError> {
-        let compiled_solc = &deployed_bytecode.metadata_hash().solc;
-        let bc_solc = &self.bc_deployed_bytecode.metadata_hash().solc;
-        if bc_solc != compiled_solc {
-            let compiled_solc = compiled_solc
-                .as_ref()
-                .map(|b| DisplayBytes::from(b.clone()).to_string());
-            let bc_solc = bc_solc
-                .as_ref()
-                .map(|b| DisplayBytes::from(b.clone()).to_string());
-            return Err(VerificationError::CompilerVersionMismatch(Mismatch::new(
-                compiled_solc,
-                bc_solc,
-            )));
+    fn compare_creation_tx_inputs(
+        remote_bytecode: &Bytecode,
+        local_bytecode: &LocalBytecode,
+    ) -> Result<(), VerificationErrorKind> {
+        let remote_creation_tx_input = remote_bytecode.creation_tx_input();
+        let local_creation_tx_input = local_bytecode.creation_tx_input();
+
+        if remote_creation_tx_input.len() < local_creation_tx_input.len() {
+            return Err(VerificationErrorKind::BytecodeLengthMismatch {
+                part: Mismatch::new(
+                    local_creation_tx_input.len(),
+                    remote_creation_tx_input.len(),
+                ),
+                raw: Mismatch::new(
+                    local_creation_tx_input.clone().into(),
+                    remote_creation_tx_input.clone().into(),
+                ),
+            });
+        }
+
+        Self::compare_bytecode_parts(
+            remote_creation_tx_input,
+            local_creation_tx_input,
+            local_bytecode.creation_tx_input_parts(),
+        )?;
+
+        Ok(())
+    }
+
+    /// Performs an actual comparison of locally compiled bytecode
+    /// with remote bytecode provided for verification.
+    ///
+    /// # Panics
+    ///
+    /// The function will panic if `remote_raw.len()` is less than `local_raw.len()`.
+    fn compare_bytecode_parts(
+        remote_raw: &Bytes,
+        local_raw: &Bytes,
+        local_parts: &Vec<BytecodePart>,
+    ) -> Result<(), VerificationErrorKind> {
+        assert!(
+            // if that fails, we would be out of range further anyway
+            remote_raw.len() >= local_raw.len(),
+            "Local bytecode is greater than remote"
+        );
+
+        let mut i = 0usize; // keep track of current processing position of `remote_raw`
+
+        for part in local_parts {
+            match part {
+                BytecodePart::Main { raw } => {
+                    if raw != &remote_raw[i..i + raw.len()] {
```

Comment:
as far as I understood, to compare this, `local_parts` has to be sorted basing on start, right? otherwise `&remote_raw[i..i + raw.len()]` will give some strange result. but I didn't find code where you sort your parts.

so is it sorted implicitly?

Reply:
Yes, we just parse the result of local compilation on Localbytecode initialization. And that parts are stored in the same order they are located in the bytecpde itself